Sources/Rendering/OpenGL/glsl/vtkVolumeFS.glsl

```diff
@@ -507,6 +510,206 @@ void getRayPointIntersectionBounds(
    check);  // 0 in 1 out
 }
 
+//=======================================================================
+// Apply the specified blend mode operation along the ray's path.
+//
+void applyBlend(vec3 posIS, vec3 stepIS, vec3 tdims, float numSteps)
+{
+  vec3 tstep = 1.0/tdims;
+
+  // integer number of steps to take and residual step size
+  int count = int(numSteps - 0.05); // end slightly inside
+  float residual = numSteps - float(count);
+
+  // local vars for the loop
+  vec4 color = vec4(0.0, 0.0, 0.0, 0.0);
+  float scalar;
+
+
+  int blendMode = //VTK::BlendMode;
+  #if vtkBlendMode == 0 // COMPOSITE_BLEND
+    for (int i = 0; i < //VTK::MaximumSamplesValue ; ++i)
+      {
+        // compute the scalar
+        vec4 tValue = getTextureValue(posIS);
+
+        // now map through opacity and color
+        vec4 tColor = getColorForValue(tValue, posIS, tstep);
+
+        float mix = (1.0 - color.a);
+
+        // this line should not be needed but nvidia seems to not handle
+        // the break correctly on windows/chrome 58 angle
+        mix = mix * sign(max(float(count - i + 1), 0.0));
+
+        color = color + vec4(tColor.rgb*tColor.a, tColor.a)*mix;
+        if (i >= count) { break; }
+        if (color.a > 0.99) { color.a = 1.0; break; }
+        posIS += stepIS;
+      }
+
+      if (color.a < 0.99)
+      {
+        posIS += (residual - 1.0)*stepIS;
+
+        // compute the scalar
+        vec4 tValue = getTextureValue(posIS);
+
+        // now map through opacity and color
+        vec4 tColor = getColorForValue(tValue, posIS, tstep);
+
+        float mix = (1.0 - color.a);
+        color = color + vec4(tColor.rgb*tColor.a, tColor.a)*mix;
+      }
+
+      gl_FragData[0] = vec4(color.rgb/color.a, color.a);
+      // gl_FragData[0] = vec4(tbounds.y/farDist, tbounds.x/farDist, color.b/color.a, 1.0);
+  #endif
+  #if vtkBlendMode == 1 || vtkBlendMode == 2
+    // MAXIMUM_INTENSITY_BLEND || MINIMUM_INTENSITY_BLEND
+    // Find maximum/minimum intensity along the ray before
+    // mapping to a color
+    float value = getTextureValue(posIS).r;
+
+    // TODO: Might be a cleaner way to write this? Can we set
+    // operator = min or operator = max instead of duplicating
+    // the code?
+    if (blendMode == 1) { // MAXIMUM_INTENSITY_BLEND
+      // Sample along the ray until MaximumSamplesValue,
+      // ending slightly inside the total distance
+      for (int i = 0; i < //VTK::MaximumSamplesValue ; ++i)
+      {
+        // compute the scalar
+        scalar = getTextureValue(posIS).r;
+
```

Comment:
Why not use tValue and tColor here? I think that would allow multiple component min/max